File: wowexportd/internal/fileio/png_writer.go
package fileio

import (
	"bytes"
	"compress/zlib"
	"encoding/binary"
	"hash/crc32"
)

// PNGWriter mirrors wow.export/src/js/png-writer.js behavior sufficiently to
// produce RGBA PNGs with adaptive per-scanline filtering and 8-bit depth.
// It does not attempt to bit-for-bit match the deflate stream, but matches
// chunk layout (IHDR, IDAT, IEND) and filtering/bit-depth settings.
type PNGWriter struct {
	Width         int
	Height        int
	bytesPerPixel int
	bitDepth      uint8
	colorType     uint8
	data          []byte
}

func NewPNGWriter(width, height int) *PNGWriter {
	return &PNGWriter{
		Width:         width,
		Height:        height,
		bytesPerPixel: 4,
		bitDepth:      8,
		colorType:     6, // RGBA
		data:          make([]byte, width*height*4),
	}
}

func (w *PNGWriter) PixelData() []byte { return w.data }

// getFilteredRaw applies adaptive filtering and returns the raw scanline buffer
// with a leading filter byte per scanline.
func (w *PNGWriter) getFilteredRaw() []byte {
	byteWidth := w.Width * w.bytesPerPixel
	raw := make([]byte, (byteWidth+1)*w.Height)

	// rolling buffer for computing sums quickly uses the destination buffer itself
	for y := 0; y < w.Height; y++ {
		dataOfs := y * byteWidth
		rawOfs := y*(byteWidth+1) + 1

		// choose best filter by sum of absolute values
		bestFilter := uint8(0)
		bestSum := int(^uint(0) >> 1) // max int
		for f := 0; f <= 4; f++ {
			sum := filterSum(byte(w.bytesPerPixel), byteWidth, y, f, w.data)
			if sum < bestSum {
				bestSum = sum
				bestFilter = uint8(f)
			}
		}
		raw[rawOfs-1] = bestFilter
		applyFilter(byte(w.bytesPerPixel), bestFilter, w.data[dataOfs:dataOfs+byteWidth], raw[rawOfs:rawOfs+byteWidth], raw[rawOfs-byteWidth-1:rawOfs-1])
	}
	return raw
}

func (w *PNGWriter) Buffer() []byte {
	raw := w.getFilteredRaw()

	// deflate raw
	var deflated bytes.Buffer
	zw := zlib.NewWriter(&deflated)
	_, _ = zw.Write(raw)
	_ = zw.Close()

	// build PNG
	var out bytes.Buffer
	// Signature
	out.Write([]byte{0x89, 'P', 'N', 'G', 0x0D, 0x0A, 0x1A, 0x0A})

	// IHDR chunk
	{
		var ihdr bytes.Buffer
		// Chunk type
		ihdr.Write([]byte{'I', 'H', 'D', 'R'})
		// Data
		var data bytes.Buffer
		binary.Write(&data, binary.BigEndian, uint32(w.Width))
		binary.Write(&data, binary.BigEndian, uint32(w.Height))
		data.WriteByte(w.bitDepth)
		data.WriteByte(w.colorType)
		data.WriteByte(0) // compression
		data.WriteByte(0) // filter
		data.WriteByte(0) // interlace

		// Length
		binary.Write(&out, binary.BigEndian, uint32(data.Len()))
		// Type + Data
		ihdr.Write(data.Bytes())
		out.Write(ihdr.Bytes())
		// CRC
		crc := crc32.ChecksumIEEE(ihdr.Bytes())
		binary.Write(&out, binary.BigEndian, crc)
	}

	// IDAT chunk
	{
		var idat bytes.Buffer
		idat.Write([]byte{'I', 'D', 'A', 'T'})
		// Length
		binary.Write(&out, binary.BigEndian, uint32(deflated.Len()))
		idat.Write(deflated.Bytes())
		out.Write(idat.Bytes())
		crc := crc32.ChecksumIEEE(idat.Bytes())
		binary.Write(&out, binary.BigEndian, crc)
	}

	// IEND chunk
	{
		binary.Write(&out, binary.BigEndian, uint32(0))
		var iend bytes.Buffer
		iend.Write([]byte{'I', 'E', 'N', 'D'})
		out.Write(iend.Bytes())
		crc := crc32.ChecksumIEEE(iend.Bytes())
		binary.Write(&out, binary.BigEndian, crc)
	}

	return out.Bytes()
}

// Filtering implementation mirrors wow.export behavior
func filterSum(bytesPerPixel byte, byteWidth int, row int, filter int, data []byte) int {
	ofs := row * byteWidth
	sum := 0
	switch filter {
	case 0: // None
		for i := 0; i < byteWidth; i++ {
			sum += absI(int(data[ofs+i]))
		}
	case 1: // Sub
		bpp := int(bytesPerPixel)
		for x := 0; x < byteWidth; x++ {
			var left byte
			if x >= bpp {
				left = data[ofs+x-bpp]
			}
			v := int(int8(data[ofs+x] - left))
			if v < 0 {
				v = -v
			}
			sum += v
		}
	case 2: // Up
		prev := ofs - byteWidth
		for x := 0; x < byteWidth; x++ {
			var up byte
			if prev >= 0 {
				up = data[prev+x]
			}
			v := int(int8(data[ofs+x] - up))
			if v < 0 {
				v = -v
			}
			sum += v
		}
	case 3: // Average
		bpp := int(bytesPerPixel)
		prev := ofs - byteWidth
		for x := 0; x < byteWidth; x++ {
			var left byte
			if x >= bpp {
				left = data[ofs+x-bpp]
			}
			var up byte
			if prev >= 0 {
				up = data[prev+x]
			}
			v := int(int8(data[ofs+x] - byte((int(left)+int(up))>>1)))
			if v < 0 {
				v = -v
			}
			sum += v
		}
	case 4: // Paeth
		bpp := int(bytesPerPixel)
		prev := ofs - byteWidth
		for x := 0; x < byteWidth; x++ {
			var left, up, upLeft byte
			if x >= bpp {
				left = data[ofs+x-bpp]
			}
			if prev >= 0 {
				up = data[prev+x]
				if x >= bpp {
					upLeft = data[prev+x-bpp]
				}
			}
			p := paeth(left, up, upLeft)
			v := int(int8(data[ofs+x] - p))
			if v < 0 {
				v = -v
			}
			sum += v
		}
	}
	return sum
}

func applyFilter(bytesPerPixel byte, filter uint8, src []byte, dst []byte, prevRow []byte) {
	byteWidth := len(src)
	switch filter {
	case 0: // None
		copy(dst, src)
	case 1: // Sub
		bpp := int(bytesPerPixel)
		for x := 0; x < byteWidth; x++ {
			var left byte
			if x >= bpp {
				left = src[x-bpp]
			}
			dst[x] = src[x] - left
		}
	case 2: // Up
		for x := 0; x < byteWidth; x++ {
			var up byte
			if len(prevRow) >= byteWidth {
				up = prevRow[x]
			}
			dst[x] = src[x] - up
		}
	case 3: // Average
		bpp := int(bytesPerPixel)
		for x := 0; x < byteWidth; x++ {
			var left byte
			if x >= bpp {
				left = src[x-bpp]
			}
			var up byte
			if len(prevRow) >= byteWidth {
				up = prevRow[x]
			}
			dst[x] = src[x] - byte((int(left)+int(up))>>1)
		}
	case 4: // Paeth
		bpp := int(bytesPerPixel)
		for x := 0; x < byteWidth; x++ {
			var left, up, upLeft byte
			if x >= bpp {
				left = src[x-bpp]
				if len(prevRow) >= byteWidth {
					upLeft = prevRow[x-bpp]
				}
			}
			if len(prevRow) >= byteWidth {
				up = prevRow[x]
			}
			dst[x] = src[x] - paeth(left, up, upLeft)
		}
	}
}

func paeth(left, up, upLeft byte) byte {
	p := int(left) + int(up) - int(upLeft)
	pLeft := absI(p - int(left))
	pUp := absI(p - int(up))
	pUpLeft := absI(p - int(upLeft))
	if pLeft <= pUp && pLeft <= pUpLeft {
		return left
	}
	if pUp <= pUpLeft {
		return up
	}
	return upLeft
}

func absI(v int) int {
	if v < 0 {
		return -v
	}
	return v
}

File: src/lib/wow/formats/png-writer.ts
/**
 * Pure-JS PNG encoder with adaptive filtering, ported byte-faithfully from
 * wow.export (src/js/png-writer.js). Output must remain byte-identical to
 * wow.export's PNG artifacts.
 */
import { BufferWrapper } from './buffer';

type FilterFn = (data: Buffer, dataOfs: number, byteWidth: number, raw: Buffer, rawOfs: number, bytesPerPixel: number) => void;
type FilterSumFn = (data: Buffer, dataOfs: number, byteWidth: number, bytesPerPixel: number) => number;

const paeth = (left: number, up: number, upLeft: number): number => {
  const p = left + up - upLeft;
  const paethLeft = Math.abs(p - left);
  const paethUp = Math.abs(p - up);
  const paethUpLeft = Math.abs(p - upLeft);

  if (paethLeft <= paethUp && paethLeft <= paethUpLeft) return left;

  if (paethUp <= paethUpLeft) return up;

  return upLeft;
};

const FILTERS: Record<number, FilterFn> = {
  // None
  0: (data, dataOfs, byteWidth, raw, rawOfs) => {
    for (let x = 0; x < byteWidth; x++) raw[rawOfs + x] = data[dataOfs + x];
  },

  // Sub
  1: (data, dataOfs, byteWidth, raw, rawOfs, bytesPerPixel) => {
    for (let x = 0; x < byteWidth; x++) {
      const left = x >= bytesPerPixel ? data[dataOfs + x - bytesPerPixel] : 0;
      raw[rawOfs + x] = data[dataOfs + x] - left;
    }
  },

  // Up
  2: (data, dataOfs, byteWidth, raw, rawOfs) => {
    for (let x = 0; x < byteWidth; x++) {
      const up = dataOfs > 0 ? data[dataOfs + x - byteWidth] : 0;
      raw[rawOfs + x] = data[dataOfs + x] - up;
    }
  },

  // Average
  3: (data, dataOfs, byteWidth, raw, rawOfs, bytesPerPixel) => {
    for (let x = 0; x < byteWidth; x++) {
      const left = x >= bytesPerPixel ? data[dataOfs + x - bytesPerPixel] : 0;
      const up = dataOfs > 0 ? data[dataOfs + x - byteWidth] : 0;
      raw[rawOfs + x] = data[dataOfs + x] - ((left + up) >> 1);
    }
  },

  // Paeth
  4: (data, dataOfs, byteWidth, raw, rawOfs, bytesPerPixel) => {
    for (let x = 0; x < byteWidth; x++) {
      const left = x >= bytesPerPixel ? data[dataOfs + x - bytesPerPixel] : 0;
      const up = dataOfs > 0 ? data[dataOfs + x - byteWidth] : 0;
      const upLeft = dataOfs > 0 && x >= bytesPerPixel ? data[dataOfs + x - (byteWidth + bytesPerPixel)] : 0;
      raw[rawOfs + x] = data[dataOfs + x] - paeth(left, up, upLeft);
    }
  },
};

const FILTER_SUMS: Record<number, FilterSumFn> = {
  // None
  0: (data, dataOfs, byteWidth) => {
    let sum = 0;
    for (let i = dataOfs, len = dataOfs + byteWidth; i < len; i++) sum += Math.abs(data[i]);

    return sum;
  },

  // Sub
  1: (data, dataOfs, byteWidth, bytesPerPixel) => {
    let sum = 0;
    for (let x = 0; x < byteWidth; x++) {
      const left = x >= bytesPerPixel ? data[dataOfs + x - bytesPerPixel] : 0;
      sum += Math.abs(data[dataOfs + x] - left);
    }

    return sum;
  },

  // Up
  2: (data, dataOfs, byteWidth) => {
    let sum = 0;
    for (let x = dataOfs, len = dataOfs + byteWidth; x < len; x++) {
      const up = dataOfs > 0 ? data[x - byteWidth] : 0;
      sum += Math.abs(data[x] - up);
    }

    return sum;
  },

  // Average
  3: (data, dataOfs, byteWidth, bytesPerPixel) => {
    let sum = 0;
    for (let x = 0; x < byteWidth; x++) {
      const left = x >= bytesPerPixel ? data[dataOfs + x - bytesPerPixel] : 0;
      const up = dataOfs > 0 ? data[dataOfs + x - byteWidth] : 0;
      sum += Math.abs(data[dataOfs + x] - ((left + up) >> 1));
    }

    return sum;
  },

  // Paeth
  4: (data, dataOfs, byteWidth, bytesPerPixel) => {
    let sum = 0;
    for (let x = 0; x < byteWidth; x++) {
      const left = x >= bytesPerPixel ? data[dataOfs + x - bytesPerPixel] : 0;
      const up = dataOfs > 0 ? data[dataOfs + x - byteWidth] : 0;
      const upLeft = dataOfs > 0 && x >= bytesPerPixel ? data[dataOfs + x - (byteWidth + bytesPerPixel)] : 0;
      sum += Math.abs(data[dataOfs + x] - paeth(left, up, upLeft));
    }

    return sum;
  },
};

/** Apply adaptive filtering to image data. */
const filter = (data: Buffer, width: number, height: number, bytesPerPixel: number): Buffer => {
  const byteWidth = width * bytesPerPixel;
  let dataOfs = 0;

  let rawOfs = 0;
  const raw = Buffer.alloc((byteWidth + 1) * height);

  let selectedFilter = 0;
  for (let y = 0; y < height; y++) {
    let min = Infinity;

    for (let i = 0, len = Object.keys(FILTERS).length; i < len; i++) {
      const sum = FILTER_SUMS[i](data, dataOfs, byteWidth, bytesPerPixel);
      if (sum < min) {
        selectedFilter = i;
        min = sum;
      }
    }

    raw[rawOfs] = selectedFilter;
    rawOfs++;

    FILTERS[selectedFilter](data, dataOfs, byteWidth, raw, rawOfs, bytesPerPixel);
    rawOfs += byteWidth;
    dataOfs += byteWidth;
  }
  return raw;
};

export class PNGWriter {
  width: number;

  height: number;

  bytesPerPixel = 4;

  bitDepth = 8;

  colorType = 6; // RGBA

  data: Buffer;

  constructor(width: number, height: number) {
    this.width = width;
    this.height = height;
    this.data = Buffer.alloc(width * height * 4);
  }

  /** Get the internal pixel data for this PNG. */
  getPixelData(): Buffer {
    return this.data;
  }

  getBuffer(): BufferWrapper {
    const filtered = new BufferWrapper(filter(this.data, this.width, this.height, this.bytesPerPixel));
    const deflated = filtered.deflate();
    const buf = BufferWrapper.alloc(8 + 25 + deflated.byteLength + 12 + 12, false);

    // 8-byte PNG signature.
    buf.writeUInt32LE(0x474E5089);
    buf.writeUInt32LE(0x0A1A0A0D);

    const ihdr = BufferWrapper.alloc(4 + 13, false);
    ihdr.writeUInt32LE(0x52444849); // IHDR
    ihdr.writeUInt32BE(this.width); // Image width
    ihdr.writeUInt32BE(this.height); // Image height
    ihdr.writeUInt8(this.bitDepth); // Bit-depth
    ihdr.writeUInt8(this.colorType); // Colour type
    ihdr.writeUInt8(0); // Compression (0)
    ihdr.writeUInt8(0); // Filter (0)
    ihdr.writeUInt8(0); // Interlace (0)
    ihdr.seek(0);

    buf.writeUInt32BE(13);
    buf.writeBuffer(ihdr);
    buf.writeInt32BE(ihdr.getCRC32());

    const idat = BufferWrapper.alloc(4 + deflated.byteLength, false);
    idat.writeUInt32LE(0x54414449); // IDAT
    idat.writeBuffer(deflated);

    idat.seek(0);

    buf.writeUInt32BE(deflated.byteLength);
    buf.writeBuffer(idat);
    buf.writeInt32BE(idat.getCRC32());

    buf.writeUInt32BE(0);
    buf.writeUInt32LE(0x444E4549); // IEND
    buf.writeUInt32LE(0x826042AE); // CRC IEND

    return buf;
  }

  /** Write this PNG to a file. */
  async write(file: string): Promise<void> {
    return this.getBuffer().writeToFile(file);
  }
}

export default PNGWriter;
